ctutil/audit.go

```diff
@@ -0,0 +1,183 @@
+//
+// This function can be used with three different types of leaf certificate:
+//   - X.509 Certificate:
+//       If using this function to verify an SCT for a normal X.509 certificate
+//       then it is enough to just provide the end entity certificate in chain.
+//       This case assumes that the SCT being provided is not embedded within
+//       the leaf certificate provided, i.e. the certificate is what was
+//       submitted to the Certificate Transparency Log in order to obtain the
+//       SCT.  For this case, set embedded to false.
+//   - Precertificate:
+//       If using this function to verify an SCT for a precertificate then the
+//       issuing certificate must also be provided in chain.  The precertificate
+//       should be at chain[0], and its issuer at chain[1].  For this case, set
+//       embedded to false.
+//   - X.509 Certificate containing the SCT embedded within it:
+//       If the SCT you wish to verify is embedded within the certificate you
+//       are providing at chain[0], set embedded to true.  VerifySCT will
+//       verify the provided SCT by building the corresponding precertificate.
+//       VerifySCT will return an error if the provided SCT cannot be found
+//       embedded within chain[0].  As with the precertificate case, the issuing
+//       certificate must also be provided in chain.  The certificate containing
+//       the embedded SCT should be at chain[0], and its issuer at chain[1].
+func VerifySCT(pubKey crypto.PublicKey, chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp, embedded bool) error {
+	leaf, err := createLeaf(chain, sct, embedded)
+	if err != nil {
+		return err
+	}
+
+	s, err := NewSignatureVerifier(pubKey)
+	if err != nil {
+		return fmt.Errorf("error creating signature verifier: %s", err)
+	}
+
+	entry := ct.LogEntry{Leaf: *leaf}
+	return s.VerifySCTSignature(*sct, entry)
+}
+
+func createLeaf(chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp, embedded bool) (*ct.MerkleTreeLeaf, error) {
+	if len(chain) == 0 {
+		return nil, errors.New("chain is empty")
+	}
+	if sct == nil {
+		return nil, errors.New("sct is nil")
+	}
+
+	sctPresent, err := ContainsSCT(chain[0], sct)
+	if err != nil {
+		return nil, fmt.Errorf("error checking for SCT in leaf certificate: %s", err)
+	}
+	if embedded && !sctPresent {
+		return nil, errors.New("SCT provided is not embedded within leaf certificate")
+	}
+
+	certType := ct.X509LogEntryType
+	if chain[0].IsPrecertificate() || embedded {
+		certType = ct.PrecertLogEntryType
+	}
+
+	var leaf *ct.MerkleTreeLeaf
+	if embedded {
+		leaf, err = MerkleTreeLeafFromChainEmbeddedSCT(chain, sct.Timestamp)
+	} else {
+		leaf, err = MerkleTreeLeafFromChain(chain, certType, sct.Timestamp)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error creating MerkleTreeLeaf: %s", err)
+	}
+	return leaf, nil
+}
+
+// ContainsSCT checks to see whether the given SCT is embedded within the given
+// certificate.
```

Comment:
Again, out of interest: do you have a use case where you get handed an embedded SCT separately from the certificate that it's embedded in?

Reply:
When writing exported functions, the only thing it is really safe to assume is that users will find new and interesting ways to use your functions, that you have not yet thought of.  When writing these functions, I assume nothing about where the things that will be passed in to the function come from, as we have no control over that.  It's all very well to say how we imagine the function being used (and to use this as justification for writing the function in the first place), but you really never know what others may come up with.  So if there is an intention as to how the function should be used, and that is the only way we want to allow it to be used, code has to be written to ensure that.

The intention with the VerifySCT function (when embedded is set to true) is exactly what you say - that the user will pass in an SCT and a certificate, where the SCT is one of the ones embedded within the certificate.  But the ContainsSCT check in VerifySCT *ensures* that this is the case.  You want to assume that a user will always loop through the embedded SCTs and only use the SCTs that are embedded in the cert to pass to this function.  The ContainsSCT check in the VerifySCT function takes that from an assumption to a guarantee.  It makes sure that users can only use the function in the way it was intended to be used.

Imagine a user doesn't read the documentation for VerifySCT, and thinks you're meant to set embedded to true whenever the certificate contains embedded SCTs.  Suppose they have submitted this final cert to a log too, and are trying to verify the SCT for the final cert, not an SCT for the corresponding precert.  So they call VerifySCT with the cert, the sct (which is not embedded in the cert), and embedded = true.  They'd get the error that the SCT does not verify, and then go on to spend a bunch of time trying to work out why.  By doing this check and returning a very clear error message stating that the SCT is not contained in the cert, this user would quickly realise the actual purpose of the embedded param, try again with it set to false, and have success.

The only real argument I can see to *not* have this check would be if we wanted it to be used to check the validity of SCTs not embedded in the cert, against the corresponding precert.  Which I don't :)

Reply:
Just a random drive-by thought:

Would it be useful to also provide some exported functions with more constrained purpose/training wheels for the common cases we know about?
e.g.:
1) I've just logged a thing, and I want to verify the SCT I got back is correct for the thing I logged
`err := audit.VerifySCT(key, loggedChain, mySCT)`
I guess I likely already have `loggedChain` prepared since I just submitted it(?)

2) I'm a CA, and I want to verify the SCTs I've just embedded into my freshly signed x.509 cert all work before I send it off.
2b) I'm writing a tool to scan CT logs and verify embedded SCTS in all the x.509 certs
`valid, invalid, err := audit.VerifyEmbeddedSCTs([]keysWotIKnow, x509Cert)`
Again I have the chain since I either just finished building the EE, or I just got it from the log

3) I'm writing a TLS client, and I want to know which and how many of the embedded SCTs are valid, and which and how many of the OSCP/TLS provided SCTs are valid
I think this is just calls to (2) followed by (1)?

Then other unexpected uses can still use the multi-ball variant if need be?

Reply:
I'd suggest a separate PR for any alternate entrypoints, given the length of this one...

For 1, just a reminder that the `LogClient.AddChain()` entrypoint will do this (if the client was configured with a public key).

(For the specific 2b) case, the extra overhead of looping over SCTs then looping over them again to check they're still there is probably overkill, and a different entrypoint would make sense)